src/gui/core/timer.cpp

```diff
@@ -133,17 +98,18 @@ size_t add_timer(const uint32_t interval,
 	DBG_GUI_E << "Adding timer.\n";
 
 	timer timer;
+	size_t next_id;
 	{
-		std::lock_guard<std::mutex> lock(timers_mutex);
+		read_lock_guard lock(timers_mutex);
```

Comment:
There is a possibility that the ID is acquired by another thread while you aren't holding either read or write lock.

You need to hold a write lock for the entire duration of this function.

Reply:
How? It's incremented before it's read, so each new thread will acquire a new value.

Reply:
Pretty much nothing in this function takes a notable amount of time. IMHO, not holding a write lock for the whole duration of this function is severe premature optimization.

Reply:
Basically, what I'm trying to do is to guess where the crash is coming from and minimize the opportunities. The crash is reported on a quite old hardware, iPad 2. So maybe it's not that premature.

I would agree to "severe" if it made the function twice as long, but it merely adds two lines and two pairs of curly brackets.

On the other hand, the crash is supposed to mean that the mutex is already being held **by the same thread**. I cannot see how `timer_callback()` can be called from other functions in this file... except for, maybe, `SDL_AddTimer()`. That's highly unlikely, but this implementation is testing that hypothesis.

Reply:
[`SDL_AddTimer()`](https://wiki.libsdl.org/SDL_AddTimer) is, in no uncertain terms, documented to run the callback in a separate thread: "The callback is run on a separate thread."

I'm afraid reducing the time this function holds the write lock looks just too dangerous to me. I'm opposed to this pull request until/unless it actually turns out to fix crashes on iOS.

Reply:
I'm of the school of thought that one should hold a lock for the least time required.

But, for clarity I would (a) refactor that period to a single code block `{...}` which might as well be a function, and I would then seriously consider using a `try ... catch ...` catching and re-throwing any exception of any type, to reduce (note: not eliminate) the chances of the lock not being released if something goes terribly wrong.

Reply:
@GregoryLundberg `std::lock_guard` is automatically released in its destructor. If you allocate it in the stack, it's guaranteed to be released when the function exits either by returning or by throwing.

Reply:
Which is fine, so long as you remember that a function can exit by neither returning nor throwing (SIGKILL for instance, among other similarly catastrophic causes) and that can lead to all sorts of hard-to-locate bugs.

Everyone knows the Three Philosophers. My favorite question from way back when I was a TA (and, yes, we dodged dinosaurs crossing the Quad): what happens if a philosopher dies while eating?

Reply:
@GregoryLundberg, I believe that when a process is killed, its mutexes are released, and if it ignores a signal, the function will resume running?
Or, how can a function exit by a signal and leave a process running at the same time?

I could only think of `longjmp`/`siglongjmp`. Hope none of Wesnoth's timer handlers calls it.

And how would a `try ... catch`, which only catches C++ exceptions, help in this scenario?

Reply:
@jyrkive Thanks for your recommendations.

A pity that nobody has come up with an idea why would `timer_callback()` crash in an attempt to double-lock a mutex.

In the light of this, I would stick to the shortest time spent in a lock.
There is no danger in not locking the whole function body; only pieces that work with shared data need to be locked.

Updating the implementation.

Reply:
I'm still opposed to the current implementation that releases the lock between allocating the timer ID and writing the ID to the `unordered_map`.

Reply:
Can you explain why it looks dangerous? All the reads and writes of a global data are protected, the generated timer ID is guaranteed to be unique.

Reply:
In particular, because it makes the "is the ID in the `unordered_map`" check useless. What if another thread adds the same ID into the map while we aren't holding the lock?

And yes, I do know that it's supposed to be impossible, as the IDs are supposed to be unique like you said. However, releasing the lock in between removes one of the guarantees that ensures unique IDs, and as a result it simply looks way too dangerous.

Reply:
I see your point.
I don't think this can happen because:
The `next_timer_id ` is currently incremented by 1 `get_next_timer_id()` whose body is a guarded block. For every ID generated, any ID generated after it will be greater (or wrapped around `size_t`).

A danger I can see here is if someone changes the code later and misses the point of locking... but this can happen to any concurrent code.

Reply:
Yeah, I understand that. But my view isn't going to change - I'm not going to accept this change to the upstream repository until/unless it's actually shown to fix crashes on iOS.

Reply:
Even given that `get_next_timer_id()` is faithful to its contract and thread safe, and is not going to wrap around in any reasonable time? I really don't understand then.

Reply:
Yes, regardless of it. I understand that the code *should* be safe even without holding the lock for the whole function, but I'm not willing to trust it.

Reply:
Okay, it will remain in in-progress iOS branch. I will see if the crashes in `timer_callback()` go, but even that is only an indirect evidence.